File: src/main/java/com/itsci/mju/maebanjumpen/controller/ReviewController.java
package com.itsci.mju.maebanjumpen.controller;

import com.itsci.mju.maebanjumpen.dto.ReviewDTO;
import com.itsci.mju.maebanjumpen.service.ReviewService;
import lombok.RequiredArgsConstructor; // ⬅️ ใช้ Lombok แทน @Autowired สำหรับ Constructor Injection
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;

@RestController
@RequestMapping("/maeban/reviews")
@RequiredArgsConstructor // ⬅️ ใช้ Constructor Injection
public class ReviewController {

    private final ReviewService reviewService;

    @GetMapping(produces = "application/json;charset=UTF-8")
    public ResponseEntity<List<ReviewDTO>> getAllReviews() {
        List<ReviewDTO> reviews = reviewService.getAllReviews();
        return ResponseEntity.ok(reviews);
    }

    @GetMapping(value = "/{id}", produces = "application/json;charset=UTF-8")
    public ResponseEntity<ReviewDTO> getReviewById(@PathVariable int id) {
        ReviewDTO review = reviewService.getReviewById(id);
        if (review == null) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.ok(review);
    }

    @GetMapping(value = "/hire/{hireId}", produces = "application/json;charset=UTF-8")
    public ResponseEntity<ReviewDTO> getReviewByHireId(@PathVariable int hireId) {
        // ⚠️ เนื่องจาก getReviewsByHireId คืนค่า List (แม้จะมีสมาชิกเดียว) เราควรใช้เมธอดที่คืนค่า ReviewDTO โดยตรง
        ReviewDTO review = reviewService.getReviewByHireId(hireId);

        if (review == null) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.ok(review);
    }

    @PostMapping(produces = "application/json;charset=UTF-8")
    public ResponseEntity<ReviewDTO> createReview(@RequestBody ReviewDTO reviewDto) { // ⬅️ รับ ReviewDTO
        try {
            ReviewDTO savedReview = reviewService.saveReview(reviewDto);
            return ResponseEntity.status(HttpStatus.CREATED).body(savedReview);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(null);
        } catch (IllegalStateException e) {
            return ResponseEntity.status(HttpStatus.CONFLICT).build();
        } catch (RuntimeException e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
        } catch (Exception e) {
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Error creating review", e);
        }
    }
    @PutMapping(value = "/{id}", produces = "application/json;charset=UTF-8")
    public ResponseEntity<ReviewDTO> updateReview(@PathVariable int id, @RequestBody ReviewDTO reviewDto) { // ⬅️ รับ ReviewDTO
        try {
            // ⬅️ เรียกใช้ Service เท่านั้น ตรรกะการอัปเดตและคำนวณเรตติ้งอยู่ใน Service
            ReviewDTO updatedReview = reviewService.updateReview(id, reviewDto);
            return ResponseEntity.ok(updatedReview);
        } catch (RuntimeException e) {
            // เช่น Review Not Found หรือ Hire Not Found
            return ResponseEntity.notFound().build();
        } catch (Exception e) {
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Error updating review", e);
        }
    }

    @DeleteMapping(value = "/{id}", produces = "application/json;charset=UTF-8")
    public ResponseEntity<Void> deleteReview(@PathVariable int id) {
        reviewService.deleteReview(id);
        return ResponseEntity.noContent().build();
    }
}

File: src/main/kotlin/com/itsci/mju/maebanjumpen/review/controller/ReviewController.kt
package com.itsci.mju.maebanjumpen.review.controller

import com.itsci.mju.maebanjumpen.review.dto.ReviewDTO
import com.itsci.mju.maebanjumpen.review.request.CreateReviewRequest
import com.itsci.mju.maebanjumpen.review.request.UpdateReviewRequest
import com.itsci.mju.maebanjumpen.review.service.ReviewService
import com.luca.intern.common.exception.BadRequestException
import com.luca.intern.common.exception.NotFoundException
import com.luca.intern.common.response.HttpResponse
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/maeban/review")
class ReviewController @Autowired internal constructor(
  private val reviewService: ReviewService
) {

  @GetMapping
  fun listAllReviews(

  ): ResponseEntity<Any> {
    return try {
      ResponseEntity.ok(
        HttpResponse(
          true,
          "รายการ review สำเร็จ",
          reviewService.listAllReviews()
        )
      )
    } catch(e: BadRequestException) {
      ResponseEntity.badRequest().body(
        HttpResponse(
          false,
          e.message ?: "รายการคำขอ review ไม่ถูกต้อง"
        )
      )
    } catch (e: NotFoundException) {
      ResponseEntity.badRequest().body(
        HttpResponse(
          false,
          e.message ?: "รายการ review ไม่พบข้อมูล"
        )
      )
    }
    catch (e: Exception) {
      ResponseEntity.badRequest().body(
        HttpResponse(
          false,
          e.message ?: "รายการ review ไม่สำเร็จ"
        )
      )
    }
  }

  @GetMapping("/{id}")
  fun getReviewById(
    @PathVariable id: Long
  ): ResponseEntity<Any> {
    return try {
      ResponseEntity.ok(
        HttpResponse(
          true,
          "รายการ review สำเร็จ",
          reviewService.getReviewById(id)
        )
      )
    } catch (e: NotFoundException) {
      ResponseEntity.badRequest().body(
        HttpResponse(
          false,
          e.message ?: "รายการ review ไม่พบข้อมูล"
        )
      )
    } catch (e: Exception) {
      ResponseEntity.badRequest().body(
        HttpResponse(
          false,
          e.message ?: "รายการ review ไม่สำเร็จ"
        )
      )
    }
  }

  @PostMapping
  fun createReview(
    @RequestBody request: CreateReviewRequest
  ): ResponseEntity<Any> {
    return try {
      ResponseEntity.ok(
        HttpResponse(
          true,
          "รายการ review สำเร็จ",
          reviewService.createReview(request)
        )
      )
    } catch (e: BadRequestException) {
      ResponseEntity.badRequest().body(
        HttpResponse(
          false,
          e.message ?: "รายการ review ไม่ต้อง"
        )
      )
    } catch (e: Exception) {
      ResponseEntity.badRequest().body(
        HttpResponse(
          false,
          e.message ?: "รายการ review ไม่สำเร็จ"
        )
      )
    }
  }

  @DeleteMapping("/{id}")
  fun deleteReview(
    @PathVariable id: Long
  ): ResponseEntity<Any> {
    return try {
      reviewService.deleteReview(id)
      ResponseEntity.ok(
        HttpResponse(
          true,
          "รายการ review สำเร็จ"
        )
      )
    } catch (e: NotFoundException) {
      ResponseEntity.badRequest().body(
        HttpResponse(
          false,
          e.message ?: "รายการ review ไม่พบข้อมูล"
        )
      )
    } catch (e: Exception) {
      ResponseEntity.badRequest().body(
        HttpResponse(
          false,
          e.message ?: "รายการ review ไม่สำเร็จ"
        )
      )
    }
  }

  @PutMapping("/{id}")
  fun updateReview(
    @PathVariable id: Long,
    @RequestBody request: UpdateReviewRequest
  ): ResponseEntity<Any> {
    return try {
      request.id = id
      ResponseEntity.ok(
        HttpResponse(
          true,
          "รายการ review สำเร็จ",
          reviewService.updateReview(request)
        )
      )
    } catch (e: NotFoundException) {
      ResponseEntity.badRequest().body(
        HttpResponse(
          false,
          e.message ?: "รายการ review ไม่พบข้อมูล"
        )
      )
    } catch (e: Exception) {
      ResponseEntity.badRequest().body(
        HttpResponse(
          false,
          e.message ?: "รายการ review ไม่สำเร็จ"
        )
      )
    }
  }

  @GetMapping("/hire/{hireId}")
  fun getReviewByHireId(
    @PathVariable hireId: Long
  ): ResponseEntity<Any> {
    return try {
      ResponseEntity.ok(
        HttpResponse(
          true,
          "รายการ review สำเร็จ",
          reviewService.getReviewByHireId(hireId)
        )
      )
    } catch (e: NotFoundException) {
      ResponseEntity.badRequest().body(
        HttpResponse(
          false,
          e.message ?: "รายการ review ไม่พบข้อมูล"
        )
      )
    } catch (e: Exception) {
      ResponseEntity.badRequest().body(
        HttpResponse(
          false,
          e.message ?: "รายการ review ไม่สำเร็จ"
        )
      )
    }
  }

  @GetMapping("/hires/{hireId}")
  fun getReviewsByHireId(
    @PathVariable hireId: Long
  ): ResponseEntity<Any> {
    return try {
      ResponseEntity.ok(
        HttpResponse(
          true,
          "รายการ review สำเร็จ",
          reviewService.getReviewsByHireId(hireId)
        )
      )
    } catch (e: Exception) {
      ResponseEntity.badRequest().body(
        HttpResponse(
          false,
          e.message ?: "รายการ review ไม่สำเร็จ"
        )
      )
    }
  }

  @GetMapping("/housekeeper/{housekeeperId}")
  fun getReviewsByHousekeeperId(
    @PathVariable housekeeperId: Long
  ): ResponseEntity<Any> {
    return try {
      ResponseEntity.ok(
        HttpResponse(
          true,
          "รายการ review สำเร็จ",
          reviewService.getReviewsByHousekeeperId(housekeeperId)
        )
      )
    } catch (e: Exception) {
      ResponseEntity.badRequest().body(
        HttpResponse(
          false,
          e.message ?: "รายการ review ไม่สำเร็จ"
        )
      )
    }
  }


}

